internal/federation/handle.go

```diff
@@ -277,3 +279,101 @@ func HandleMediaRequests(mediaIds map[string]func(w http.ResponseWriter)) func(*
 		mediamux.Handle("/r0/download/{origin}/{mediaId}", downloadFn).Methods("GET")
 	}
 }
+
+// HandleTransactionRequests is an option which will process GET /_matrix/federation/v1/send/{transactionID} requests universally when requested.
+func HandleTransactionRequests() func(*Server) {
+	return func(srv *Server) {
+		srv.mux.Handle("/_matrix/federation/v1/send/{transactionID}", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			// Check federation signature
+			fedReq, errResp := gomatrixserverlib.VerifyHTTPRequest(
+				req, time.Now(), gomatrixserverlib.ServerName(srv.ServerName), srv.keyRing,
+			)
+			if fedReq == nil {
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Read the request body
+			body, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				errResp := util.MessageResponse(400, err.Error())
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Unmarshal the request body into a transaction object
+			var transaction gomatrixserverlib.Transaction
+			err = json.Unmarshal(body, &transaction)
+			if err != nil {
+				errResp := util.MessageResponse(400, err.Error())
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Extract the transaction ID from the request vars
+			vars := mux.Vars(req)
+			transaction.TransactionID = gomatrixserverlib.TransactionID(vars["transactionID"])
+
+			// Transactions are limited in size; they can have at most 50 PDUs and 100 EDUs.
+			// https://matrix.org/docs/spec/server_server/latest#transactions
+			if len(transaction.PDUs) > 50 || len(transaction.EDUs) > 100 {
+				errResp := util.MessageResponse(400, "Transactions are limited to 50 PDUs and 100 EDUs")
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Construct a response and fill as we process each PDU
+			response := gomatrixserverlib.RespSend{}
+			response.PDUs = make(map[string]gomatrixserverlib.PDUResult, 0)
+			for _, pdu := range transaction.PDUs {
+				var header struct {
+					RoomID string `json:"room_id"`
+				}
+				if err := json.Unmarshal(pdu, &header); err != nil {
+					log.Printf("Transaction '%s': Failed to extract room ID from event", transaction.TransactionID)
+					// We don't know the event ID at this point so we can't return the
+					// failure in the PDU results
+				}
+
+				// Retrieve the room version from the server
+				room := srv.rooms[header.RoomID]
```

Comment:
Need to handle if `room == nil`

Reply:
Will do. Could you explain why a room would be nil if we have the ID for it?

Reply:
Complement should not assume that HSes are going to send valid traffic to it. The room ID is untrusted input (from a random HS), so it's entirely possible that it is referring to a room which Complement is unaware of.